### PR TITLE
feat: add `is_power_of_ten` function and tests

### LINKED ISSUE #94: [Feature]: `is_power_of_ten`
### What is the feature you would like to see?

Follow the [reference](https://github.com/MystenLabs/deepbookv3/blob/main/packages/deepbook/sources/helper/math.move#L89-L101).

### PR DESCRIPTION
**Pull request by Rajesh-Royal**
Resolves #94 

#### PR Checklist

- [x] Tests
- [x] Documentation
- [ ] Changelog

### Overview
- Added `is_power_of_ten` utility for u64 type, allowing users to check if a given value is a power of ten (e.g., 1, 10, 100).
- Comprehensive test coverage for `is_power_of_ten`, including positive, negative, and edge cases to ensure correctness and reliability.

## Summary by CodeRabbit

* **New Features**
  * Added power-of-ten validation across integer types (u8, u16, u32, u64, u128, u256), letting you check whether a value is an exact power of ten.

* **Tests**
  * Added extensive unit tests covering basic cases, edge cases, large magnitudes, and binary-search lookup paths to validate correctness.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by bidzyyys — CHANGES_REQUESTED** [review]
Hey @Rajesh-Royal!
Please add `is_power_of_ten` to all the primitive data types:
- [x] `u8`
- [x] `u16`
- [x] `u32`
- [x] `u64`
- [x] `u128`
- [x] `u256`

I converted the PR to draft state.

File: math/core/sources/u64.move [hunk 1]
```diff
@@ -108,3 +108,16 @@ public fun inv_mod(value: u64, modulus: u64): Option<u64> {
...
+public fun is_power_of_ten(n: u64): bool {
+    let mut num = n;
+    if (num < 1) {
+        false
+    } else {
+        while (num % 10 == 0) {
+            num = num / 10;
+        };
+        num == 1
+    }
```
[the remark this answers is not in the document]
Comment (Rajesh-Royal) [reply math/core/sources/u64.move]: @ericnordelo  To optimize `is_power_of_ten` lookups (faster than division/log math for fixed small sets), I'm using precomputed constant checks for `u32` and binary search on a lookup table for larger types like `u128`.

For `u32` powers: 10^0 to 10^9:

```move
/// Returns true if the value is a power of ten (1, 10, 100, ...)
/// 
/// For u32, valid powers of ten are: 1, 10, 100, ..., 10^9 (1000000000)
public fun is_power_of_ten(n: u32): bool {
    n == 1 || n == 10 || n == 100 || n == 1000 || n == 10000 ||
    n == 100000 || n == 1000000 || n == 10000000 || n == 100000000 || n == 1000000000
}
```

For `u128` powers: 10^0 to 10^38:

```move
/// Returns true if the value is a power of ten (1, 10, 100, ...).
/// Uses binary search on a lookup table for efficiency.
/// For u128, valid powers: 10^0 (1) to 10^38.
public fun is_power_of_ten(n: u128): bool {
    // Precomputed powers of 10: 10^0 to 10^38
   let powers = vector[
        1u128,
        10u128,
        100u128,
        ...remaining
    ];
    
   // Binary search
    let mut left = 0;
    let mut right = powers.length();
    
    while (left < right) {
        let mid = left + (right - left) / 2;
        let mid_val = *powers.borrow(mid);
        
        if (mid_val == n) {
            return true
        } else if (mid_val < n) {
            left = mid + 1;
        } else {
            right = mid;
        }
    };
    
    false
}
```

This avoids runtime math ops (repeated division by 10) for gas efficiency on-chain. Thoughts, or should I explore bit tricks instead?

**Comment by Rajesh-Royal** [comment]
@ericnordelo, @bidzyyys,  @immrsd  Is there anything else needed from my side to get this PR merged, or is it good to go?

File: math/core/sources/u128.move [hunk 2]
```diff
@@ -107,3 +107,71 @@ public fun inv_mod(value: u128, modulus: u128): Option<u128> {
...
+        100000000000000000000000000000000u128,
+        1000000000000000000000000000000000u128,
+        10000000000000000000000000000000000u128,
+        100000000000000000000000000000000000u128,
+        1000000000000000000000000000000000000u128,
+        10000000000000000000000000000000000000u128,
+        100000000000000000000000000000000000000u128,
+    ];
+
+    // Binary search
```
[the remark this answers is not in the document]
Comment (bidzyyys) [reply math/core/sources/u128.move]: Like in this file: https://github.com/OpenZeppelin/contracts-sui/blob/main/math/core/sources/internal/macros.move

**Comment by ericnordelo** [comment]
I understand benchmarking takes time, and since the PR looks solid, maybe we can merge it and we will finish polishing these details on future iterations before the end of the milestone. What do you think @bidzyyys

**Comment by bidzyyys** [comment]
> I understand benchmarking takes time, and since the PR looks solid, maybe we can merge it and we will finish polishing these details on future iterations before the end of the milestone. What do you think @bidzyyys

Only [this comment](https://github.com/OpenZeppelin/contracts-sui/pull/125#discussion_r2643145820) and I am happy to merge.

**Comment by Rajesh-Royal** [comment]
> > I haven’t checked the gas usage yet, but assuming the calculation is optimized, it should be gas-efficient too.
> 
> It would be great if you can confirm that.
> 
> Besides that, you can format files with something like `prettier-move "**/*.move" -w`

Would it make sense to add this formatting command to the documentation for .move files? `npx -y @mysten/prettier-plugin-move "**/*.move" -w`

File: math/core/sources/u128.move (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rajesh-Royal) [reply math/core/sources/u128.move]: > The code for binary search in `u128` & `u256` looks like the same. IMHO it should be refactor as a helper macro that works for each data type.

Please refer to this commit - https://github.com/OpenZeppelin/contracts-sui/pull/125/commits/a069d528560f5fa655f06b7ec708473333f9008e

File: math/core/sources/u64.move [hunk 3]
```diff
@@ -108,3 +108,16 @@ public fun inv_mod(value: u64, modulus: u64): Option<u64> {
 public fun mul_mod(a: u64, b: u64, modulus: u64): u64 {
     macros::mul_mod!(a, b, modulus)
 }
+
+/// Returns true if the value is a power of ten (1, 10, 100, ...)
+public fun is_power_of_ten(n: u64): bool {
```
[the remark this answers is not in the document]
Comment (bidzyyys) [reply math/core/sources/u64.move]: @Rajesh-Royal This solution is still needed.

**Comment by bidzyyys** [comment]
@CodeRabbit review.
